drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -0,0 +1,819 @@
+
+	if (!ep_state || !USB_EP_DIR_IS_OUT(ep)) {
+		LOG_ERR("Not valid endpoint: %02x", ep);
+		return -EINVAL;
+	}
+
+	size_t bytes_received =
+		(state.setup_available ? SETUP_SIZE : udc_rpi_get_ep_buffer_len(ep));
+
+	state.setup_available = false;
+
+	/* If no more data in the buffer, start a new read transaction. */
+	LOG_DBG("received %d offset: %d", bytes_received, ep_state->read_offset);
+	if (bytes_received == ep_state->read_offset) {
+		return usb_dc_ep_start_read(ep, DATA_BUFFER_SIZE);
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_read(const uint8_t ep, uint8_t *const data, const uint32_t max_data_len,
+		   uint32_t *const read_bytes)
+{
+	if (usb_dc_ep_read_wait(ep, data, max_data_len, read_bytes) != 0) {
+		return -EINVAL;
+	}
+
+	if (!max_data_len) {
+		return 0;
+	}
+
+	if (usb_dc_ep_read_continue(ep) != 0) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_halt(const uint8_t ep)
+{
+	return usb_dc_ep_set_stall(ep);
+}
+
+int usb_dc_ep_flush(const uint8_t ep)
+{
+	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
+
+	if (!ep_state) {
+		return -EINVAL;
+	}
+
+	LOG_ERR("Not implemented");
+
+	return 0;
+}
+
+int usb_dc_ep_mps(const uint8_t ep)
+{
+	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
+
+	if (!ep_state) {
+		return -EINVAL;
+	}
+
+	return ep_state->mps;
+}
+
+int usb_dc_detach(void)
+{
+	LOG_ERR("Not implemented");
+
+	return 0;
+}
+
+int usb_dc_reset(void)
+{
+	LOG_ERR("Not implemented");
+
+	return 0;
+}
+
+/*
+ * This thread is only used to not run the USB device stack and endpoint
+ * callbacks in the ISR context, which happens when an callback function
+ * is called. TODO: something similar should be implemented in the USB
+ * device stack so that it can be used by all drivers.
+ */
+static void udc_rpi_thread_main(void *arg1, void *unused1, void *unused2)
+{
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+	struct cb_msg msg;
+
+	while (true) {
+		k_msgq_get(&usb_dc_msgq, &msg, K_FOREVER);
+
+		if (msg.type == USB_DC_CB_TYPE_EP) {
+			struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(msg.ep);
+
+			switch (msg.cb) {
+			case USB_DC_EP_SETUP:
+				if (ep_state->cb) {
+					ep_state->cb(msg.ep,
+						USB_DC_EP_SETUP);
+				}
+				break;
+			case USB_DC_EP_DATA_OUT:
+				if (ep_state->cb) {
+					ep_state->cb(msg.ep,
+						USB_DC_EP_DATA_OUT);
+				}
+				break;
+			case USB_DC_EP_DATA_IN:
+				if (ep_state->cb) {
+					ep_state->cb(msg.ep,
+						USB_DC_EP_DATA_IN);
+				}
+				break;
+			default:
+				LOG_ERR("unknown msg");
+				break;
+			}
+		} else if (state.status_cb) {
+			switch (msg.cb) {
+			case USB_DC_RESET:
+				state.status_cb(USB_DC_RESET, NULL);
+				break;
+			case USB_DC_ERROR:
+				state.status_cb(USB_DC_ERROR, NULL);
+				break;
+			case USB_DC_SUSPEND:
+				state.status_cb(USB_DC_SUSPEND, NULL);
+				break;
+			case USB_DC_RESUME:
+				state.status_cb(USB_DC_RESUME, NULL);
+				break;
+			default:
+				LOG_ERR("unknown msg");
+				break;
+			}
```

Comment:
```diff c
diff --git a/drivers/usb/device/usb_dc_rpi_pico.c b/drivers/usb/device/usb_dc_rpi_pico.c
index 84bc8f0cbea..48577e668be 100644
--- a/drivers/usb/device/usb_dc_rpi_pico.c
+++ b/drivers/usb/device/usb_dc_rpi_pico.c
@@ -793,26 +793,12 @@ static void udc_rpi_thread_main(void *arg1, void *unused1, void *unused2)
                                }
                                break;
                        default:
-                               LOG_ERR("unknown msg");
+                               LOG_ERR("unknown ep msg");
                                break;
                        }
-               } else if (state.status_cb) {
-                       switch (msg.cb) {
-                       case USB_DC_RESET:
-                               state.status_cb(USB_DC_RESET, NULL);
-                               break;
-                       case USB_DC_ERROR:
-                               state.status_cb(USB_DC_ERROR, NULL);
-                               break;
-                       case USB_DC_SUSPEND:
-                               state.status_cb(USB_DC_SUSPEND, NULL);
-                               break;
-                       case USB_DC_RESUME:
-                               state.status_cb(USB_DC_RESUME, NULL);
-                               break;
-                       default:
-                               LOG_ERR("unknown msg");
-                               break;
+               } else {
+                       if (state.status_cb) {
+                               state.status_cb(msg.cb, NULL);
                        }
                }
        }
```

Reply:
Refactored this to simplify greatly. Leaving open until you review again.